### PR TITLE
gpt-auto: reference dissected partitions to mount via diskseq block device symlinks

### PR DESCRIPTION
**Pull request by poettering**
Let's tighten the rules a bit, so that replacing the root disk while booting isn't as likely a to trigger a vulnerability as it is right now.

### CODE REVIEW

File: src/shared/dissect-image.c [hunk 1]
```diff
@@ -290,28 +290,57 @@ static void dissected_partition_done(DissectedPartition *p) {
...
+                bool need_p;
 
-        if (isempty(whole_devname)) /* Make sure there *is* a last char */
-                return -EINVAL;
+                /* Given a whole block device node name (e.g. /dev/sda or /dev/loop7) generate a partition
+                 * device name (e.g. /dev/sda7 or /dev/loop7p5). The rule the kernel uses is simple: if whole
+                 * block device node name ends in a digit, then suffix a 'p', followed by the partition
+                 * number. Otherwise, just suffix the partition number without any 'p'. */
+
+                if (nr < 0) { /* no partition? */
```
Comment (yuwata) [inline src/shared/dissect-image.c]: This should be `nr <= 0`, or return `-EINVAL` when `nr == 0`, as partition number should be always positive.

File: src/shared/dissect-image.c [hunk 2]
```diff
@@ -290,28 +290,57 @@ static void dissected_partition_done(DissectedPartition *p) {
...
+                        s = strdup(whole_devname);
+                        if (!s)
+                                return -ENOMEM;
+                } else {
+
+                        if (isempty(whole_devname)) /* Make sure there *is* a last char */
```
Comment (yuwata) [inline src/shared/dissect-image.c]: This should be checked also when `nr <= 0`.

**Review by yuwata — APPROVED** [review]
LGTM. Two minor comments.

**Review by yuwata — CHANGES_REQUESTED** [review]
Hmm, I noticed several issues:
- this does not check `ID_IGNORE_DISKSEQ` udev property,
- when gpt-auto is executed, udevd may not processed the device yet, hence ID_IGNORE_DISKSEQ may not set yet. If the udev property set after gpt-auto finished, then the generated .mount units will fail.

**Comment by yuwata** [comment]
How about to introduce say `DiskSeq=UINT64` setting in `[Mount]` section, and check the diskseq before/after mount?

**Comment by poettering** [comment]
> Hmm, I noticed several issues:
> 
>     * this does not check `ID_IGNORE_DISKSEQ` udev property,
> 
>     * when gpt-auto is executed, udevd may not processed the device yet, hence ID_IGNORE_DISKSEQ may not set yet. If the udev property set after gpt-auto finished, then the generated .mount units will fail.

gpt-auto will only run for bog standard disks where the ESP was found on and that have a properly set up GPT and so on, i.e. nothing ID_IGNORE_DISKSEQ would ever be set for (i.e. MD or so). Hence this should not be an issue.

i.e. if you have complex storage, then gpt-auto doesn't enable it itself anyway, and thus I don#t think we need to care.

**Comment by poettering** [comment]
> How about to introduce say `DiskSeq=UINT64` setting in `[Mount]` section, and check the diskseq before/after mount?

I want something like that, but I don#t think thta should be in systemd. See →  https://github.com/util-linux/util-linux/issues/1786

So yeah, once that is implemented in util-linux I think we should start generating suitable mount options to lock this down fully.

but I think that can happen at a later stage.

File: src/shared/dissect-image.c (hunk 2 above)
Comment (poettering) [reply src/shared/dissect-image.c]: well, this is just an underflow check for the calculation one line below.

i'll rework to make this clearer.

**Comment by poettering** [comment]
Force pushed new version with the two items addressed. PTAL

**Comment by yuwata** [comment]
> > Hmm, I noticed several issues:
> > ```
> > * this does not check `ID_IGNORE_DISKSEQ` udev property,
> > 
> > * when gpt-auto is executed, udevd may not processed the device yet, hence ID_IGNORE_DISKSEQ may not set yet. If the udev property set after gpt-auto finished, then the generated .mount units will fail.
> > ```
> 
> gpt-auto will only run for bog standard disks where the ESP was found on and that have a properly set up GPT and so on, i.e. nothing ID_IGNORE_DISKSEQ would ever be set for (i.e. MD or so). Hence this should not be an issue.
> 
> i.e. if you have complex storage, then gpt-auto doesn't enable it itself anyway, and thus I don#t think we need to care.

I see. Thanks. Then, LGTM.

File: src/shared/dissect-image.c [hunk 3]
```diff
@@ -322,28 +322,60 @@ static void dissected_partition_done(DissectedPartition *p) {
...
+                 * device name (e.g. /dev/sda7 or /dev/loop7p5). The rule the kernel uses is simple: if whole
+                 * block device node name ends in a digit, then suffix a 'p', followed by the partition
+                 * number. Otherwise, just suffix the partition number without any 'p'. */
+
+                if (nr < 0) { /* whole disk? */
+                        s = strdup(whole_devname);
+                        if (!s)
+                                return -ENOMEM;
+                } else {
+                        assert(nr != 0); /* zero is not a valid partition nr */
```
Comment (yuwata) [inline src/shared/dissect-image.c]: I think it is better to move this at the beginning of the function.

File: src/shared/dissect-image.c [hunk 4]
```diff
@@ -322,28 +322,60 @@ static void dissected_partition_done(DissectedPartition *p) {
...
+
+                        size_t l = strlen(whole_devname);
+                        if (l < 1) /* underflow check for the subtraction below */
+                                return -EINVAL;
+                        need_p = ascii_isdigit(whole_devname[l-1]); /* Last char a digit? */
```
Comment (yuwata) [inline src/shared/dissect-image.c]: `need_p` can be declared here.

**Review by yuwata — APPROVED** [review]

**Pull request merged**